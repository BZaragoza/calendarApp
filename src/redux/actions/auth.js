import Swal from 'sweetalert2';

import { fetchWithoutToken, fetchWithToken } from '../../helpers/fetch';
import { types } from '../types/index';

export const startLogin = (email, password) => {
  return async dispatch => {
    
    const res = await fetchWithoutToken('auth', {email, password}, 'POST')
    const body = await res.json();

    if (body.ok) {
      localStorage.setItem('token', body.token)
      localStorage.setItem('tokenCreation', +new Date());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    } else {
      Swal.fire('Error', body.msg, 'error')
    }
  }
}

export const login = (user) => ({
  type: types.authLogin,
  payload: user
})


export const startRegister = (email, password, name) => {
  return async dispatch => {

    const res = await fetchWithoutToken('auth/new', {name, email, password}, 'POST');
    const body = await res.json();

    if (body.ok) {
      localStorage.setItem('token', body.token)
      localStorage.setItem('tokenCreation', +new Date());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    } else {
      return Swal.fire('Error', body.msg, 'error')
    }

  }
}

export const startChecking = () => {
  return async dispatch => {

    const res = await fetchWithToken('auth/renew');
    const body = await res.json();

    if (body.ok) {
      localStorage.setItem('token', body.token)
      localStorage.setItem('tokenCreation', +new Date());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    } else {
      dispatch(checkingFinish())
    }

  }
}

const checkingFinish = () => ({
  type: types.authCheckingFinished,
})

export const startLogout = () => {
  return dispatch => {
    localStorage.removeItem('token')
    localStorage.removeItem('tokenCreation')

    dispatch(logout())
  }
}

export const logout = () => ({
  type: types.authLogout
})