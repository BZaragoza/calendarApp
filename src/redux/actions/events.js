import Swal from "sweetalert2";

import { types } from "../types";

import {fetchWithToken} from '../../helpers/fetch'
import { manipulateEvents } from '../../helpers/manipulateEvents';

export const startEventSetEvents = () => {
  return async dispatch => {
    const res = await fetchWithToken('events');
    const body = await res.json();

    const events = manipulateEvents(body?.events);

    dispatch(eventSetEvents(events))
  }
}

const eventSetEvents = (events) => ({
  type: types.eventSetEvents,
  payload: events
});

export const startEventAddNew = (event) => {
  return async (dispatch, getState) => {

    const {uid, name} = getState().auth;

   try {
    const res = await fetchWithToken('events', event, 'POST');
    const body = await res.json();

    if (body.ok) {
      event.id = body.event.id;
      event.user = {
        _id: uid,
        name
      }

      dispatch(eventAddNew(event))
    }

   } catch (error) {
    console.log(error)
   }
  }
}

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
})

export const startEventUpdate = (event) => {
  return async dispatch => {

    try {

      const res = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await res.json();

      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        return Swal.fire('Error', body.msg, 'error')
      }


    } catch (error) {
      console.log(error)
    }

  }
}

const eventUpdated = (event) => ({
  type: types.eventUpdate,
  payload: event,
})

export const startEventDelete = () => {
  return async (dispatch, getState) => {

    const { activeEvent } = getState().calendar;

    try {
      const res = await fetchWithToken(`events/${activeEvent.id}`, {}, 'DELETE');
      const body = await res.json();

      if (body.ok) {
        dispatch(eventDeleted())
      } else {
        Swal.fire('Error', body.msg, 'error');
      }
    } catch (error) {
      console.log(error);
    }

  }
}

const eventDeleted = () => ({
  type: types.eventDeleted,
})

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
})  

export const eventClearActive = () => ({
  type: types.eventClearActive
})  

export const eventClearLogout = () => ({
  type: types.eventLogout
})