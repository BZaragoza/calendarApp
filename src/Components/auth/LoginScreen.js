import Swal from 'sweetalert2';
import { useDispatch } from 'react-redux';

import useForm from '../../hooks/useForm';
import { startLogin, startRegister } from '../../redux/actions/auth';

import "./login.css";


const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formLoginValues, handleLoginInputChange ] = useForm({
    loginEmail: '',
    loginPassword: '',
  });
  
  const [ formRegisterValues, handleRegiserInputChange ] = useForm({
    registerName: 'Esteban',
    registerEmail: 'esteban@zaragoza.com',
    // registerPassword: '123456',
    // registerPassword1: '123456',
  });
  
  const {loginEmail, loginPassword} = formLoginValues;
  const {registerName, registerEmail, registerPassword, registerPassword1} = formRegisterValues;


  const handleLogin = (e) => {
    e.preventDefault();

    dispatch(startLogin(loginEmail, loginPassword))
  }

  const handleRegister = (e) => {
    e.preventDefault()

    if (registerPassword !== registerPassword1) {
      return Swal.fire('Error', 'Passwords doesn\'t match', 'error')
    }

    dispatch(startRegister(registerEmail, registerPassword, registerName));
  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Email"
                name="loginEmail"
                value={loginEmail}
                onChange={handleLoginInputChange}  
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name="loginPassword"
                value={loginPassword}
                onChange={handleLoginInputChange}  
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Name"
                name="registerName"
                value={registerName}
                onChange={handleRegiserInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Email"
                name="registerEmail"
                value={registerEmail}
                onChange={handleRegiserInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name="registerPassword"
                value={registerPassword}
                onChange={handleRegiserInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Confirm password"
                name="registerPassword1"
                value={registerPassword1}
                onChange={handleRegiserInputChange}
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Create acount" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

export default LoginScreen;