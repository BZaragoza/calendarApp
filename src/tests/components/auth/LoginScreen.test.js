import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux';
import { mount } from 'enzyme'
import thunk from 'redux-thunk';

import LoginScreen from '../../../Components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../redux/actions/auth';
import Swal from 'sweetalert2';

Swal.fire = jest.fn();
jest.mock('../../../redux/actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn(),
}))

const mockStore = configureStore([thunk]);
const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
)

describe('Tests on <LoginScreen />', () => {

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('Should call login\'s dispatch', () => {
    wrapper.find('input[name="loginEmail"]').simulate('change', {
      target: {
        name: 'loginEmail',
        value: 'brian@zaragoza.com'
      }
    })
    wrapper.find('input[name="loginPassword"]').simulate('change', {
      target: {
        name: 'loginPassword',
        value: '123456'
      }
    })

    wrapper.find('form').at(0).prop('onSubmit')({
      preventDefault(){}
    })

    expect(startLogin).toHaveBeenCalledWith('brian@zaragoza.com', '123456');
  })

  test('Shouldn\'t register if passwords don\'t match', () => {
    wrapper.find('input[name="registerPassword"]').simulate('change', {
      target: {
        name: 'registerPassword',
        value: '123456'
      }
    })
    wrapper.find('input[name="registerPassword1"]').simulate('change', {
      target: {
        name: 'registerPassword1',
        value: '654321'
      }
    })

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });


    expect(startRegister).not.toHaveBeenCalledWith("esteban@gmail.com", "123456", "Esteban")
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Passwords doesn\'t match', 'error')

  })

  test('Should register with matching passwords', () => {
    wrapper.find('input[name="registerPassword"]').simulate('change', {
      target: {
        name: 'registerPassword',
        value: '123456'
      }
    })
    wrapper.find('input[name="registerPassword1"]').simulate('change', {
      target: {
        name: 'registerPassword1',
        value: '123456'
      }
    })

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });


    expect(Swal.fire).not.toHaveBeenCalledWith('Error', 'Passwords doesn\'t match', 'error')
    expect(startRegister).toHaveBeenCalledWith("esteban@zaragoza.com", "123456", "Esteban")

  })

})