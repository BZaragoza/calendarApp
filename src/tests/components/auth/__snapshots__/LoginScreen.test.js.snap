// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <LoginScreen /> Should render correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value="Esteban"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value="esteban@zaragoza.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword1"
            onChange={[Function]}
            placeholder="Confirm password"
            type="password"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create acount"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
